core/include/usBundleContext.h

```diff
@@ -691,7 +736,18 @@ class US_Core_EXPORT BundleContext
   friend class BundlePrivate;
   template<class S> friend struct ServiceHolder;
 
+  // allow templated code to use the internal logger
+  template<class S, class TTT, class R> friend class BundleAbstractTracked;
+  template<class S, class T> friend class ServiceTracker;
+  template<class S, class TTT> friend class ServiceTrackerPrivate;
+  template<class S, class TTT> friend class TrackedService;
+
   BundleContext(BundlePrivate* bundle);
+
+  // Not for use by clients of the Framework.
+  // Provides access to the Framework's log sink to allow templated code
+  // to log diagnostic information. 
+  std::shared_ptr<LogSink> GetLogSink();
```

Comment:
I started working on this and am realizing that this may be more trouble than its worth. 

I can't see a way to push ```GetLogSink()``` into ```BundleContextPrivate``` without exposing some other private header/function. Even if I was to export ```GetPrivate()``` from the library (like what is done with ```MakeBundleContext()``` in ```usGetBundleContext.h```) clients can then access the ```BundleContextPrivate``` object by doing ```GetPrivate(MakeBundleContext(...))```.

I'm not sure this is worth the effort given our timeline for 3.0 right now. 
Is there a better solution which doesn't expose private implementation classes?